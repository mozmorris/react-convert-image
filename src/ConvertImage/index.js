/**
 *
 * ConvertImage module
 *
 */

// Vendor
import React, { Component } from 'react';
import PropTypes from 'prop-types';

class ConvertImage extends Component {
  static propTypes = {
    image: PropTypes.string.isRequired,
    format: PropTypes.oneOf(['webp', 'jpeg', 'png']),
    onConversion: PropTypes.func.isRequired,
  };

  static defaultProps = {
    format: 'webp',
  };

  state = {
    convertedImage: undefined,
  };

  componentDidMount() {
    this.node.addEventListener('load', this.convert);
  }

  shouldComponentUpdate(nextProps) {
    return nextProps.image !== this.state.convertedImage;
  }

  setRef = (node) => {
    this.node = node;
  };

  convert = () => {
    const { props: { onConversion, format } } = this;
    const canvas = document.createElement('canvas');
    const ctx = canvas.getContext('2d');

    canvas.width = this.node.naturalWidth;
    canvas.height = this.node.naturalHeight;
    ctx.drawImage(this.node, 0, 0);

    const convertedImage = canvas.toDataURL(`image/${format}`);

    this.setState({ convertedImage }, () => onConversion(convertedImage));
  };

  render() {
    const { props: { image } } = this;

    return (
      <img ref={this.setRef} src={image} />
    );
  }
}

export default ConvertImage;
