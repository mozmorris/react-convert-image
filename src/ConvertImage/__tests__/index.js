/**
 *
 * Test for ConvertImage component
 *
 */

// Vendor
import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

// App
import ConvertImage from '../';

const mockProps = {
  image: 'data:image/bmp;base64,abc123',
  onConversion() {},
};

// Setup component
const setup = (props = {}, method = mount) => {
  return method(
    <ConvertImage
      {...mockProps}
      {...props}
    />,
  );
};

describe('ConvertImage', () => {
  const getContext = window.HTMLCanvasElement.prototype.getContext;
  const drawImage = jest.fn();
  const getContextMock = () => ({ drawImage });

  beforeEach(() => {
    window.HTMLCanvasElement.prototype.getContext = getContextMock;
  });

  afterEach(() => {
    drawImage.mockReset();
    window.HTMLCanvasElement.prototype.getContext = getContext;
  });

  describe('render', () => {
    it('should render the', () => {
      const component = setup();

      expect(toJson(component)).toMatchSnapshot();
    });
  });

  describe('onConversion', () => {
    it('should be called', () => {
      const onConversion = jest.fn();

      const component = setup({ onConversion });

      component.instance().convert();

      expect(onConversion).toHaveBeenCalled();
    });
  });
});
