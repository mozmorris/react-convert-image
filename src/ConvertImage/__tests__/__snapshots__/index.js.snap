// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConvertImage render should render the 1`] = `
<ConvertImage
  format="webp"
  image="data:image/bmp;base64,abc123"
  onConversion={[Function]}
>
  <img
    src="data:image/bmp;base64,abc123"
  />
</ConvertImage>
`;
